### PR TITLE
Add `FragmentInfo`.

### PR DESCRIPTION
**Pull request by teo-tsirpanis**
This PR implements TileDB's fragment info API, and adds tests to it, ported from Java.

[SC-22321](https://app.shortcut.com/tiledb-inc/story/22321/add-high-level-wrapper-for-fragmentinfo-api)

### CODE REVIEW

File: sources/TileDB.CSharp/Marshalling/ScratchBuffer.cs
Comment (teo-tsirpanis) [inline sources/TileDB.CSharp/Marshalling/ScratchBuffer.cs:16]: Copied from https://gist.github.com/teo-tsirpanis/db9ae014efcad311a5df87e6ed04bc51. I can strip license headers, it's mine and in public domain.

File: tests/TileDB.CSharp.Test/FragmentInfoTest.cs
Comment (ihnorton) [inline tests/TileDB.CSharp.Test/FragmentInfoTest.cs:177]: Can we parametrize this over different numeric types?

Comment (teo-tsirpanis) [reply tests/TileDB.CSharp.Test/FragmentInfoTest.cs:177]: I copy-pasted the MBR test to work with attributes with 32-bit integers. We could avoid duplication with [.NET 7's generic math](https://learn.microsoft.com/en-us/dotnet/standard/generics/math) but until then…

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (teo-tsirpanis) [inline sources/TileDB.CSharp/FragmentInfo.cs:414]: Thoughts about this? To check the generic type we would have to allocate an `ArraySchema`; the function currently does not allocate. Is it OK to return garbage if the user specifies the wrong type? This is an advanced API either way.

Comment (teo-tsirpanis) [reply sources/TileDB.CSharp/FragmentInfo.cs]: Done.

File: sources/TileDB.CSharp/Enums.cs
Comment (teo-tsirpanis) [inline sources/TileDB.CSharp/Enums.cs:1166]: Thoughts on this? Returning `Any` on an unsupported type can be misleading since `Any` is obsolete and there is no clear documentation on what to do with it.

File: sources/TileDB.CSharp/Marshalling/MarshaledStringOut.cs
Comment (teo-tsirpanis) [inline sources/TileDB.CSharp/Marshalling/MarshaledStringOut.cs:22]: I didn't handle UCS-2 and UCS-4; they are legacy, don't have framework support and need custom code to convert. Is that OK?

**Comment by teo-tsirpanis** [comment]
[CI is displaying these errors.](https://github.com/TileDB-Inc/TileDB-CSharp/actions/runs/3523622203/jobs/5908294088#step:8:43)

File: sources/TileDB.CSharp/Marshalling/MarshaledStringOut.cs
Comment (ihnorton) [reply sources/TileDB.CSharp/Marshalling/MarshaledStringOut.cs:22]: Yes

File: sources/TileDB.CSharp/Enums.cs
Comment (ihnorton) [reply sources/TileDB.CSharp/Enums.cs:1166]: What is `default` here?

Comment (ihnorton) [reply sources/TileDB.CSharp/Enums.cs:1166]: I think throwing is fine, if you mean `TILEDB_ANY`, there's definitely no way to translate that.

Comment (teo-tsirpanis) [reply sources/TileDB.CSharp/Enums.cs:1166]: In general `default` is equal to `null` on reference types, zero on primitives and all fields recursively set to `default` in value types.

In this case I added it to satisfy the compiler. It won't be returned because the above function will throw (we do it in separate method for performance reasons).

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (teo-tsirpanis) [inline sources/TileDB.CSharp/FragmentInfo.cs:60]: This API is obsolete. The keys are now supposed to be set through the `Config`, right?

File: sources/TileDB.CSharp/Enums.cs
Comment (teo-tsirpanis) [inline sources/TileDB.CSharp/Enums.cs:1107]: I don't like this, it restricts us to a one-to-one mapping between .NET types and TileDB data types (it will be a problem once we allow mapping between TileDB and .NET booleans). We might need a `bool IsCompatible<T>(DataType dataType)` method.

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (ihnorton) [inline sources/TileDB.CSharp/FragmentInfo.cs:490]: Does this implicitly mean that the maximum size is 128? Or will ScratchBuffer spill to the heap if larger?

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (ihnorton) [inline sources/TileDB.CSharp/FragmentInfo.cs:481]: These should probably be restricted to TILEDB_STRING_ASCII for now.

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (ihnorton) [reply sources/TileDB.CSharp/FragmentInfo.cs:490]: Ok, I see how it works - nice 💯

**Review by ihnorton — APPROVED** [review]
One minor nit, otherwise LGTM - please feel free to merge when that one is addressed, or skip if not applicable.

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (teo-tsirpanis) [reply sources/TileDB.CSharp/FragmentInfo.cs:466]: Why?

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (teo-tsirpanis) [reply sources/TileDB.CSharp/FragmentInfo.cs]: Removed it after seeing the deprecation policy.

File: sources/TileDB.CSharp/FragmentInfo.cs
Comment (ihnorton) [reply sources/TileDB.CSharp/FragmentInfo.cs:466]: That's the only supported var-size type for dimensions. We _should_ throw an error from the C API, but you could give a better/earlier error here. Doesn't need to block merging though.

Comment (teo-tsirpanis) [reply sources/TileDB.CSharp/FragmentInfo.cs:466]: If there is no chance for the C API to give us a Unicode string-typed dimension, the relevant decoding code will just not get executed. I will keep it since it might be useful for other string decoding cases.

**Pull request merged**
